### DISCUSSION TITLE
What is the recommended way of having multiple async dependencies in a FutureProvider/StreamProvider

### DISCUSSION THREAD

**Discussion by timcreatedit** [rrousselGit/riverpod#2062] [Q&A] [answered]
In our app we have multiple Future/StreamProviders that depend on other asyncronous providers.
So far we have used this basic solution for that:

```dart
final resultProvider = FutureProvider((ref) async {
  final a = await ref.watch(aProvider.future);
  final b = await ref.watch(bProvider.future);
  final c = await ref.watch(cProvider.future);
  return ResultingClass(a, b, c);
});
```

This can however lead to the exception described in #815.
The solution proposed there would mean writing:

```dart 
final resultProvider = FutureProvider((ref) async {
  final a = ref.watch(aProvider.future);
  final b = ref.watch(bProvider.future);
  final c = ref.watch(cProvider.future);
  return ResultingClass(
    await a, 
    await b,
    await c,
  );
});
```

That works for some cases, however sometimes one of the FutureProviders depends on the result of another which makes this impossible

```dart
final resultProvider = FutureProvider((ref) async {
  final a = ref.watch(aProvider.future);
  final b = ref.watch(bProvider.future);
  final c = ref.watch(cProvider(b).future);
  return ResultingClass(
    await a, 
    await b,
    await c,
  );
});
```

We're wondering what the best practice here is. We see a few options:
- Just handling the Exception, the functionality seems to work
- Avoiding provider.future and using AsyncValue instead, however this introduces more nesting and logic
- ...?

Thanks for your help!

**Comment by rrousselGit** [comment, marked as answer]
The idea is to delay the "await" as much as possible. If you need to call "await", so be it.

To begin with, the mentioned issue is something that's going to change in the future anyway. I plan on updating the logic around this

**Reply by timcreatedit** [reply]
That is really nice to hear! I've always thought this part to be the most unintuitive thing in riverpod, together with FutureProvider.autoDispose not allowing `await ref.read(futureProvider.future)`. Glad to hear both of these are being adressed :) Thank you!

**Reply by rrousselGit** [reply]
> Together with FutureProvider.autoDispose not allowing await ref.read(futureProvider.future)

It should technically work now. But the problem is `read`.
`read` is flawed and one of my goals is to find ways to remove it.
